module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/**
+ * Secure cookie permission provider for VuFind.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Catalyst IT 2024.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Authorization
+ * @author   Alex Buckley <alexbuckley@catalyst.net.nz>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     http://www.vufind.org  Main Page
+ */
+
+namespace VuFind\Role\PermissionProvider;
+
+use VuFind\Cookie\CookieManager;
+
+/**
+ * Secure cookie permission provider for VuFind.
+ *
+ * @category VuFind
+ * @package  Authorization
+ * @author   Alex Buckley <alexbuckley@catalyst.net.nz>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+class SecureCookie implements PermissionProviderInterface
+{
+    use \VuFind\Log\LoggerAwareTrait;
+
+    /**
+     * Constructor
+     *
+     * @param CookieManager $cookieManager Cookie manager
+     */
+    public function __construct(
+        protected CookieManager $cookieManager
+    ) {
```

Comment:
My suggestion would be to add a second constructor parameter that's a `\Laminas\Session\Container` object (you can find many examples of building containers elsewhere in the code if you need them -- note that it's important in the factory to inject the session manager into the container when you create it, so you should copy this logic from another factory that creates containers). The container namespace could be something like 'SecureCookie'.

My thought is that when you set a cookie designed to authorize the user with this plugin, you set the cookie to a randomly-generated value like a uuid and then also store that value in the session container. Then when the cookie is validated, we confirm that the value in the session matches the value in the cookie, preventing remote spoofing since there is no way to predict the value of the cookie. We could create a public helper method on the provider to set the secret, and then any code setting an authorization cookie could fetch the provider out of the plugin manager and store the secret so that it will work as expected on subsequent requests.

Does that make any sense?

Reply:
Thanks @demiankatz ! Yes, I think so.  I've been working on a follow-up adding this, I will push it to this PR in the next day or two.

Reply:
thanks @demiankatz 

I've added a follow-up which hopefully does the following correctly:

- Add a second constructor parameter to SecureCookie.php which is a \Laminas\Session\Container object 

- In the SecureCookieFactory.php inject the session manager into the container when we create it

Reply:
Hi @demiankatz I have added the public helper method in the SecureCookie.php to set the secret (a random uuid) in the cookie and session container. 

I'm also validating in the SecureCookie.php if the value in the session matches the value in the cookie.

I'm not sure if I've done this right, so if you could you please advise what I should improve upon that would be much appreciated!